lib/Dialect/Comb/CombFolds.cpp

```diff
@@ -758,6 +758,40 @@ OpFoldResult AndOp::fold(ArrayRef<Attribute> constants) {
   return constFoldAssociativeOp(constants, hw::PEO::And);
 }
 
+/// returns a single common operand that all inputs of the operation `op` can
+/// be traced back to, or an empty `Value` if no such operand exists.
+///
+/// for example for `or(a[0], a[1], ..., a[n-1])` this function returns `a`
+/// (assuming the bit-width of `a` is `n`)
+template <typename Op>
+static Value getCommonOperand(Op op) {
+  auto inputs = op.inputs();
+
+  auto commonOperand = [&inputs](size_t index) -> Value {
+    if (auto extractOp = inputs[index].template getDefiningOp<ExtractOp>()) {
+      return extractOp.getOperand();
+    } else {
+      return Value();
+    }
+  };
+
+  Value source = commonOperand(0);
+
+  size_t size = inputs.size();
+
+  if (!source || size != source.getType().getIntOrFloatBitWidth()) {
```

Comment:
I think `size != source.getType().getIntOrFloatBitWidth()` is not sufficient condition for exhaustiveness check in the case where operands are not CSEd. In the example below, `%0` and `%1` are semantically equivalent but different values. The current implementation incorrectly canonicalizes the example into `%b == -1`. It would be necessary to check extracted bits are actually exhaustive.
```mlir
hw.module @reductionAndOps(%a: i1, %b: i2) -> (c: i1) {
  %0 = comb.extract %b from 0 : (i2) -> i1
  %1 = comb.extract %b from 0 : (i2) -> i1
  %2 = comb.and %0, %1 : i1
  hw.output %2 : i1
}
```